File: app/src/main/java/com/topnotch/developers/view/MainActivity.java
package com.topnotch.developers.view;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.TelephonyManager;
import android.view.Menu;
import android.view.View;

import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.test.espresso.IdlingResource;

import com.topnotch.developers.R;
import com.topnotch.developers.adapter.GithubAdapter;
import com.topnotch.developers.databinding.ActivityMainBinding;
import com.topnotch.developers.dialogs.DialogInternetCheck;
import com.topnotch.developers.interfaces.EndlessRecyclerOnScrollListener;
import com.topnotch.developers.interfaces.IGithubUsersView;
import com.topnotch.developers.interfaces.IOnFilterListener;
import com.topnotch.developers.interfaces.OnInternetRetryListener;
import com.topnotch.developers.model.GithubUser;
import com.topnotch.developers.presenter.GithubPresenter;
import com.topnotch.developers.util.EspressoIdlingResource;
import com.topnotch.developers.util.NetworkUtility;

import java.util.List;
import java.util.Locale;

public class MainActivity extends AppCompatActivity implements IGithubUsersView, SwipeRefreshLayout.OnRefreshListener, OnInternetRetryListener, View.OnClickListener, IOnFilterListener {
    private GithubAdapter adapter;
    private GithubPresenter presenter;
    private NetworkUtility networkUtility = new NetworkUtility();
    private String query;
    private ActivityMainBinding binding;
    private int page = 1;
    private boolean isLoading = false;
    private boolean isLastPage = false;
    private GridLayoutManager gridLayoutManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);
        query = "location:" + getCountry();
        loadGithubUsers(query, page);

        binding.swipeToReflesh.setOnRefreshListener(this);
        adapter = new GithubAdapter(this, this);
        gridLayoutManager = new GridLayoutManager(this, 2);
        binding.recyclerView.setLayoutManager(gridLayoutManager);
        binding.recyclerView.setAdapter(adapter);
        binding.recyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener(gridLayoutManager) {
            @Override
            public void onScrolledToEnd() {
                if (!isLoading && !isLastPage) {
                    page++;
                    isLoading = true;
                    loadGithubUsers(query, page);
                }
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public void githubReadyUsers(List<GithubUser> users) {
        isLoading = false;
        binding.swipeToReflesh.setRefreshing(false);
        binding.recyclerView.setVisibility(View.VISIBLE);
        adapter.updateUsers(users);
        EspressoIdlingResource.decrement();
    }

    @Override
    public void displayUser(GithubUser githubUser) {
        Intent intent = new Intent(this, DetailActivity.class);
        intent.putExtra("username", githubUser.getLogin());
        intent.putExtra("imageUrl", githubUser.getAvatarUrl());
        startActivity(intent);
    }

    @Override
    public void onRefresh() {
        loadGithubUsers(query, page);
    }

    public void loadGithubUsers(String query, int page) {
        if (networkUtility.networkAvailable(this)) {
            binding.swipeToReflesh.setRefreshing(true);
            int per_page = 100;
            presenter = new GithubPresenter(this, query, page, per_page);
            EspressoIdlingResource.increment();
        } else {
            showInternetDialog();
        }
    }

    public void searchDeveloper(String q) {
        isLoading = true;
        if (networkUtility.networkAvailable(this)) {
            presenter = new GithubPresenter(this, q, 0, 1);
            EspressoIdlingResource.increment();
        } else {
            showInternetDialog();
        }
    }

    private void showInternetDialog() {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        DialogInternetCheck dialogInternetCheck = new DialogInternetCheck();
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
        fragmentTransaction.add(android.R.id.content, dialogInternetCheck, "Internet Dialog").addToBackStack(null).commit();
    }

    public String getCountry() {
        try {
            TelephonyManager tm = (TelephonyManager) this.getSystemService(Context.TELEPHONY_SERVICE);
            Locale locale = new Locale("", tm.getNetworkCountryIso().toUpperCase());
            return locale.getDisplayCountry();
        } catch (Exception ex) {
            return Locale.getDefault().getDisplayCountry();
        }
    }

    @VisibleForTesting
    public IdlingResource getCountingIdlingResource() {
        return EspressoIdlingResource.getIdlingResource();
    }

    //Activate search menu
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.search_menu, menu);
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                searchDeveloper(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                adapter.getFilter().filter(query);
                return false;
            }
        });
        return true;
    }

    @Override
    public void onRetry(boolean retry) {
        loadGithubUsers(query, page);
    }

    @Override
    public void onClick(View v) {
        this.presenter.displayClickedUser((GithubUser) v.getTag());
    }

    @Override
    public void onFilterListener(String keyWord, Boolean isFound) {
        if (isFound) {
            binding.recyclerView.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/com/topnotch/developers/view/MainActivity.kt
package com.topnotch.developers.view

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.telephony.TelephonyManager
import android.view.Menu
import android.view.View
import androidx.annotation.VisibleForTesting
import androidx.appcompat.widget.SearchView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.test.espresso.IdlingResource
import com.topnotch.developers.R
import com.topnotch.developers.adapter.GithubAdapter
import com.topnotch.developers.databinding.ActivityMainBinding
import com.topnotch.developers.dialogs.DialogNoRecord
import com.topnotch.developers.interfaces.*
import com.topnotch.developers.model.GithubUser
import com.topnotch.developers.presenterImpl.GithubUsersPresenterImpl
import com.topnotch.developers.service.GetUsersNoticeIntractorImpl
import com.topnotch.developers.util.EspressoIdlingResource
import java.util.*


class MainActivity : BaseActivity(), OnInternetRetryListener, RecyclerItemClickListener,
    IMainContract.usersListingView, SwipeRefreshLayout.OnRefreshListener, IOnRecordNotFound {
    private lateinit var binding: ActivityMainBinding
    private var adapter: GithubAdapter? = null
    private var page = 1
    private val PAGE_LIMIT = 20
    private var isLoading = false
    private var isLastPage = false
    private lateinit var presenter: IMainContract.usersPresenter
    private var query = ""
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        initRecyclerView()
        initProgressBar()
        query = "location:$country"
        presenter = GithubUsersPresenterImpl(this, GetUsersNoticeIntractorImpl())
        presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
        binding.swipeRefresh.setOnRefreshListener(this)
    }

    private fun initRecyclerView() {
        val gridLayoutManager = GridLayoutManager(this, 2)
        adapter = GithubAdapter(this)
        binding.recyclerView.adapter = adapter
        binding.recyclerView.layoutManager = GridLayoutManager(this, 2)
        binding.recyclerView.addOnScrollListener(object :
            EndlessRecyclerOnScrollListener(gridLayoutManager) {
            override fun onScrolledToEnd() {
                if (!isLoading && !isLastPage) {
                    page += 1
                    presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
                }
            }

        })
    }

    val country: String
        get() = try {
            val tm = this.getSystemService(TELEPHONY_SERVICE) as TelephonyManager
            val locale = Locale("", tm.networkCountryIso.uppercase())
            locale.displayCountry
        } catch (ex: Exception) {
            Locale.getDefault().displayCountry
        }

    @get:VisibleForTesting
    val countingIdlingResource: IdlingResource
        get() = EspressoIdlingResource.getIdlingResource()

    //Activate search menu
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.search_menu, menu)
        val searchManager = getSystemService(SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.action_search).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.maxWidth = Int.MAX_VALUE
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
                return false
            }

            override fun onQueryTextChange(query: String): Boolean {
//                adapter.getFilter().filter(query)
                return false
            }
        })
        return true
    }

    override fun onRetry(retry: Boolean) {
        binding.swipeRefresh.isRefreshing = false
        presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
    }

    override fun onItemClicked(githubUser: GithubUser) {
        val intent = Intent(this, DetailActivity::class.java)
        intent.putExtra("username", githubUser.login)
        intent.putExtra("imageUrl", githubUser.avatarUrl)
        startActivity(intent)
    }

    override fun showProgress() {
        isLoading = true
        progressBar.visibility = View.VISIBLE
    }

    override fun hideProgress() {
        isLoading = false
        progressBar.visibility = View.GONE
    }

    override fun setDataToRecyclerView(githubUsers: List<GithubUser>) {
        if (githubUsers.isNotEmpty()) {
            adapter!!.addUsers(githubUsers)
        } else {
            showDialogFragment(DialogNoRecord(this, query), getString(R.string.no_record_found))
        }
        isLastPage = githubUsers.size < PAGE_LIMIT
    }

    override fun onResponseFailure(error: Any) {
        errorHandler(error)
    }

    override fun onRefresh() {
        progressBar.visibility = View.VISIBLE
        presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
    }

    override fun onRecordNotFound(boolean: Boolean?) {
        query = "location:$country"
        presenter.requestUsersDataFromServer(query, page, PAGE_LIMIT)
    }
}